py/sentry_streams/flink/flink_fn_translator.py

```diff
@@ -0,0 +1,133 @@
+from datetime import timedelta
+from typing import Any, Callable, Generic
+
+from pyflink.common import Time, TypeInformation, Types
+from pyflink.datastream.functions import AggregateFunction, KeySelector
+from pyflink.datastream.window import (
+    CountSlidingWindowAssigner,
+    CountTumblingWindowAssigner,
+    SlidingEventTimeWindows,
+    TumblingEventTimeWindows,
+    WindowAssigner,
+)
+
+from sentry_streams.user_functions.function_template import (
+    Accumulator,
+    GroupBy,
+    InputType,
+    IntermediateType,
+    OutputType,
+)
+from sentry_streams.window import (
+    MeasurementUnit,
+    SlidingCountWindow,
+    SlidingEventTimeWindow,
+    TumblingCountWindow,
+    TumblingEventTimeWindow,
+    Window,
+)
+
+FLINK_TYPE_MAP: dict[Any, Any] = {
+    tuple[str, int]: (Types.TUPLE, [Types.STRING, Types.INT]),
+    str: (Types.STRING, []),
+}
+
+
+def convert_to_flink_type(type: Any) -> TypeInformation:
+    """
+    Convert Python types to Flink's Types class.
+    Recommended for use in explicitly defining the
+    output_type of a Flink operator.
+    """
+    fn: Callable[..., TypeInformation]
+    fn, args = FLINK_TYPE_MAP[type]
+    flink_type = fn([arg() for arg in args]) if args else fn()
+
+    return flink_type
+
+
+class FlinkAggregate(AggregateFunction, Generic[InputType, IntermediateType, OutputType]):
+    """
+    Takes the Streams API's Accumulator and transforms it into Flink's
+    Aggregate Function. For a streaming pipeline runing on Flink,
+    this is the main mechanism for an Accumulator-based aggregation
+    on a window of data. Adds incoming data to an intermediate
+    aggregate and merges intermediate aggregates.
+    """
+
+    def __init__(self, acc: Accumulator[InputType, IntermediateType, OutputType]) -> None:
+        self.acc = acc
+
+    def create_accumulator(self) -> IntermediateType:
+        return self.acc.create()
+
+    def add(self, value: InputType, accumulator: IntermediateType) -> IntermediateType:
+        return self.acc.add(accumulator, value)
+
+    def get_result(self, accumulator: IntermediateType) -> OutputType:
+        return self.acc.get_output(accumulator)
+
+    def merge(
+        self,
+        acc_a: IntermediateType,
+        acc_b: IntermediateType,
+    ) -> IntermediateType:
+        return self.acc.merge(acc_a, acc_b)
```

Comment:
I am not sure I actually get the goal of this class.
This class is purely a wrapper around the `Accumulator` you pass in the initializer. It does not seem to add anything. 
- Every method call just routes the same call to the Accumulator class without adding any additional parameter. It seems to me the `Accumulator[InputType, IntermediateType, OutputType]` class should fit everywhere you use this class without adding this abstraction. 
- Am I missing anything ?

Is it here only to decouple Flink from our API further.

Reply:
Flink expects an `AggregateFunction` for the aggregation on windows. We, in the Streams API, want to provide a standard template for an Accumulator that users will implements. 

We will translate it to various runtime-specific interfaces. This is the Flink interface. 

Does that make sense? I also added a docstring to explain it.